src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Value/UriConverterFactory.cs

```diff
@@ -0,0 +1,22 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+namespace System.Text.Json.Serialization.Converters
+{
+    /// <summary>
+    /// Avoid loading System.Private.Uri.dll until necessary.
+    /// </summary>
+    internal sealed class UriConverterFactory : JsonConverterFactory
+    {
+        public override bool CanConvert(Type type)
+        {
+            return type.ToString() == "System.Uri";
```

Comment:
What will happen if somebody has their own `System.Uri` type?

Reply:
As written, this will probably lead to an InvalidOperationException thrown [here](https://github.com/dotnet/runtime/blob/6b2f3b5ef717ed21b5fbeaae034e135bdf044f90/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs#L203). This input is likely malicious, and the exception thrown here may be sufficient.

We should probably return `false` here for this input, maybe by doing:

```csharp
return type.ToString() == "System.Uri" && Type.GetType("System.Uri, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089") == type;
```

We do this when detecting non-generic `Stack` & `Queue` types: https://github.com/dotnet/runtime/blob/6b2f3b5ef717ed21b5fbeaae034e135bdf044f90/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs#L255

Not sure if this goes ahead to load System.Private.Uri.dll, but maybe that's okay if the input type's name is `System.Uri`.

Either way, we should add a test case for this.

Reply:
As written, this will probably lead to an InvalidOperationException thrown [here](https://github.com/dotnet/runtime/blob/6b2f3b5ef717ed21b5fbeaae034e135bdf044f90/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Converters.cs#L203). This input is likely malicious, and the exception thrown here may be sufficient.

We should probably return `false` here for this input, maybe by doing:

```csharp
return type.ToString() == "System.Uri" && Type.GetType("System.Uri, System, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089") == type;
```

We do this when detecting non-generic `Stack` & `Queue` types: https://github.com/dotnet/runtime/blob/6b2f3b5ef717ed21b5fbeaae034e135bdf044f90/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/Collection/IEnumerableConverterFactoryHelpers.cs#L255

Not sure if this goes ahead to load System.Private.Uri.dll, but maybe that's okay if the input type's name is `System.Uri`.

Either way, we should add a test case for this.

Reply:
Yes we should have a common pattern for deferred loading of types and also loading types where we don't have an assembly reference. I'll look at this a bit and get back.

Reply:
Just make sure that what you come up with is compatible with IL linker :-)

Reply:
> Either way, we should add a test case for this (if we can).

Adding a "simple" test is not possible since that add a "System.Uri" type that collides the real type. Not worth it IMO. If it is deemed important then we should also add tests for the other deferred types that we don't have an assembly reference to.